### ISSUE TITLE
Clangd doesn't work in files without an extension

### ISSUE THREAD

**Issue by bstaletic** [clangd/clangd#16]
Given a file named `no_extension` with the contents of `int x = 5;`, upon `didOpen` there's an error saying `Could not build CompilerInvocation for file /home/bstaletic/Temp/sparsehash-c11/no_extension`. `hover` response is `{"error":{"code":-32001,"message":"invalid AST"},"id":"2","jsonrpc":"2.0"}` and `completion` response is `{"id":"3","jsonrpc":"2.0","result":{"isIncomplete":false,"items":[]}}`.

The complete clangd log:

```
V[21:59:09.337] <<< {"id":"1","jsonrpc":"2.0","method":"initialize","params":{"capabilities":{"textDocument":{"completion":{"completionItemKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]}}}},"initializationOptions":{},"processId":4848,"rootPath":"/home/bstaletic/Temp/sparsehash-c11","rootUri":"file:///home/bstaletic/Temp/sparsehash-c11"}}

I[21:59:09.337] <-- initialize("1")
I[21:59:09.337] --> reply:initialize("1") 0 ms
V[21:59:09.337] >>> {"id":"1","jsonrpc":"2.0","result":{"capabilities":{"codeActionProvider":true,"completionProvider":{"resolveProvider":false,"triggerCharacters":[".",">",":"]},"declarationProvider":true,"definitionProvider":true,"documentFormattingProvider":true,"documentHighlightProvider":true,"documentOnTypeFormattingProvider":{"firstTriggerCharacter":"}","moreTriggerCharacter":[]},"documentRangeFormattingProvider":true,"documentSymbolProvider":true,"executeCommandProvider":{"commands":["clangd.applyFix","clangd.applyTweak"]},"hoverProvider":true,"referencesProvider":true,"renameProvider":true,"signatureHelpProvider":{"triggerCharacters":["(",","]},"textDocumentSync":2,"workspaceSymbolProvider":true}}}

V[21:59:09.338] <<< {"jsonrpc":"2.0","method":"initialized","params":{}}

I[21:59:09.338] <-- initialized
I[21:59:09.338] unhandled notification initialized
V[21:59:09.338] <<< {"jsonrpc":"2.0","method":"workspace/didChangeConfiguration","params":{"settings":{}}}

I[21:59:09.338] <-- workspace/didChangeConfiguration
V[21:59:09.339] <<< {"jsonrpc":"2.0","method":"textDocument/didOpen","params":{"textDocument":{"languageId":"cpp","text":"auto x = 5;\n","uri":"file:///home/bstaletic/Temp/sparsehash-c11/foo","version":1}}}

I[21:59:09.339] <-- textDocument/didOpen
I[21:59:09.340] Updating file /home/bstaletic/Temp/sparsehash-c11/foo with command [/home/bstaletic/Temp/sparsehash-c11/build/tests] /usr/sbin/clang++ -I/home/bstaletic/Temp/sparsehash-c11/. -I/home/bstaletic/Temp/sparsehash-c11/tests/. -Wall -Wextra -Wpedantic -Wno-missing-field-initializers -O3 -DNDEBUG -O3 -c /home/bstaletic/Temp/sparsehash-c11/foo -resource-dir=/usr/lib/clang/9.0.0
V[21:59:09.343] Ignored diagnostic. /home/bstaletic/Temp/sparsehash-c11/foo: 'linker' input unused
V[21:59:09.343] Ignored diagnostic. argument unused during compilation: '-I /home/bstaletic/Temp/sparsehash-c11/.'
V[21:59:09.343] Ignored diagnostic. argument unused during compilation: '-I /home/bstaletic/Temp/sparsehash-c11/tests/.'
V[21:59:09.343] Ignored diagnostic. argument unused during compilation: '-Wall'
V[21:59:09.343] Ignored diagnostic. argument unused during compilation: '-Wextra'
V[21:59:09.343] Ignored diagnostic. argument unused during compilation: '-Wpedantic'
V[21:59:09.343] Ignored diagnostic. argument unused during compilation: '-Wno-missing-field-initializers'
V[21:59:09.343] Ignored diagnostic. argument unused during compilation: '-O3'
V[21:59:09.343] Ignored diagnostic. argument unused during compilation: '-D NDEBUG'
V[21:59:09.343] Ignored diagnostic. argument unused during compilation: '-O3'
V[21:59:09.343] Ignored diagnostic. argument unused during compilation: '-c'
V[21:59:09.343] Ignored diagnostic. unable to handle compilation, expected exactly one compiler job in ''
E[21:59:09.343] Could not build CompilerInvocation for file /home/bstaletic/Temp/sparsehash-c11/foo
V[21:59:13.216] <<< {"id":"2","jsonrpc":"2.0","method":"textDocument/hover","params":{"position":{"character":5,"line":0},"textDocument":{"uri":"file:///home/bstaletic/Temp/sparsehash-c11/foo"}}}

I[21:59:13.217] <-- textDocument/hover("2")
V[21:59:13.220] Ignored diagnostic. /home/bstaletic/Temp/sparsehash-c11/foo: 'linker' input unused
V[21:59:13.220] Ignored diagnostic. argument unused during compilation: '-I /home/bstaletic/Temp/sparsehash-c11/.'
V[21:59:13.220] Ignored diagnostic. argument unused during compilation: '-I /home/bstaletic/Temp/sparsehash-c11/tests/.'
V[21:59:13.220] Ignored diagnostic. argument unused during compilation: '-Wall'
V[21:59:13.220] Ignored diagnostic. argument unused during compilation: '-Wextra'
V[21:59:13.221] Ignored diagnostic. argument unused during compilation: '-Wpedantic'
V[21:59:13.221] Ignored diagnostic. argument unused during compilation: '-Wno-missing-field-initializers'
V[21:59:13.221] Ignored diagnostic. argument unused during compilation: '-O3'
V[21:59:13.221] Ignored diagnostic. argument unused during compilation: '-D NDEBUG'
V[21:59:13.221] Ignored diagnostic. argument unused during compilation: '-O3'
V[21:59:13.221] Ignored diagnostic. argument unused during compilation: '-c'
V[21:59:13.221] Ignored diagnostic. unable to handle compilation, expected exactly one compiler job in ''
I[21:59:13.221] --> reply:textDocument/hover("2") 4 ms, error: invalid AST
V[21:59:13.221] >>> {"error":{"code":-32001,"message":"invalid AST"},"id":"2","jsonrpc":"2.0"}

V[22:10:25.890] <<< {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"text":"auto x = 5;\n\n"}],"textDocument":{"uri":"file:///home/bstaletic/Temp/sparsehash-c11/foo","version":2}}}

I[22:10:25.890] <-- textDocument/didChange
V[22:10:25.892] <<< {"id":"3","jsonrpc":"2.0","method":"textDocument/completion","params":{"position":{"character":0,"line":1},"textDocument":{"uri":"file:///home/bstaletic/Temp/sparsehash-c11/foo"}}}

I[22:10:25.892] <-- textDocument/completion("3")
V[22:10:25.896] Ignored diagnostic. /home/bstaletic/Temp/sparsehash-c11/foo: 'linker' input unused
V[22:10:25.896] Ignored diagnostic. argument unused during compilation: '-I /home/bstaletic/Temp/sparsehash-c11/.'
V[22:10:25.896] Ignored diagnostic. argument unused during compilation: '-I /home/bstaletic/Temp/sparsehash-c11/tests/.'
V[22:10:25.896] Ignored diagnostic. argument unused during compilation: '-Wall'
V[22:10:25.896] Ignored diagnostic. argument unused during compilation: '-Wextra'
V[22:10:25.896] Ignored diagnostic. argument unused during compilation: '-Wpedantic'
V[22:10:25.896] Ignored diagnostic. argument unused during compilation: '-Wno-missing-field-initializers'
V[22:10:25.896] Ignored diagnostic. argument unused during compilation: '-O3'
V[22:10:25.896] Ignored diagnostic. argument unused during compilation: '-D NDEBUG'
V[22:10:25.896] Ignored diagnostic. argument unused during compilation: '-O3'
V[22:10:25.896] Ignored diagnostic. argument unused during compilation: '-c'
V[22:10:25.896] Ignored diagnostic. unable to handle compilation, expected exactly one compiler job in ''
E[22:10:25.896] Couldn't create CompilerInvocation
I[22:10:25.896] Code complete: 0 results from Sema, 0 from Index, 0 matched, 0 returned.
I[22:10:25.896] --> reply:textDocument/completion("3") 4 ms
V[22:10:25.896] >>> {"id":"3","jsonrpc":"2.0","result":{"isIncomplete":false,"items":[]}}

I[22:10:26.392] Updating file /home/bstaletic/Temp/sparsehash-c11/foo with command [/home/bstaletic/Temp/sparsehash-c11/build/tests] /usr/sbin/clang++ -I/home/bstaletic/Temp/sparsehash-c11/. -I/home/bstaletic/Temp/sparsehash-c11/tests/. -Wall -Wextra -Wpedantic -Wno-missing-field-initializers -O3 -DNDEBUG -O3 -c /home/bstaletic/Temp/sparsehash-c11/foo -resource-dir=/usr/lib/clang/9.0.0
V[22:10:26.396] Ignored diagnostic. /home/bstaletic/Temp/sparsehash-c11/foo: 'linker' input unused
V[22:10:26.396] Ignored diagnostic. argument unused during compilation: '-I /home/bstaletic/Temp/sparsehash-c11/.'
V[22:10:26.396] Ignored diagnostic. argument unused during compilation: '-I /home/bstaletic/Temp/sparsehash-c11/tests/.'
V[22:10:26.396] Ignored diagnostic. argument unused during compilation: '-Wall'
V[22:10:26.396] Ignored diagnostic. argument unused during compilation: '-Wextra'
V[22:10:26.396] Ignored diagnostic. argument unused during compilation: '-Wpedantic'
V[22:10:26.396] Ignored diagnostic. argument unused during compilation: '-Wno-missing-field-initializers'
V[22:10:26.396] Ignored diagnostic. argument unused during compilation: '-O3'
V[22:10:26.396] Ignored diagnostic. argument unused during compilation: '-D NDEBUG'
V[22:10:26.396] Ignored diagnostic. argument unused during compilation: '-O3'
V[22:10:26.396] Ignored diagnostic. argument unused during compilation: '-c'
V[22:10:26.396] Ignored diagnostic. unable to handle compilation, expected exactly one compiler job in ''
E[22:10:26.396] Could not build CompilerInvocation for file /home/bstaletic/Temp/sparsehash-c11/foo
```

Note: Providing a different `compile_commands.json` or even removing one completely doesn't change the behaviour at all.

**Comment by kadircet** [comment]
It looks like inferred compile commands is missing the language
option(-xc++/header) for files without extensions

Could you also provide compile commands.json?

**Comment by bstaletic** [comment]
```
[
{
  "directory": "/home/bstaletic/Temp/sparsehash-c11/build/tests",
  "command": "/usr/sbin/clang++   -I/home/bstaletic/Temp/sparsehash-c11/. -I/home/bstaletic/Temp/sparsehash-c11/tests/.   -std=c++11 -Wall -Wextra -Wpedantic -Wno-missing-field-initializers -O3 -DNDEBUG -O3   -o CMakeFiles/bench.dir/bench.cc.o -c /home/bstaletic/Temp/sparsehash-c11/tests/bench.cc",
  "file": "/home/bstaletic/Temp/sparsehash-c11/tests/bench.cc"
},
{
  "directory": "/home/bstaletic/Temp/sparsehash-c11/build/tests",
  "command": "/usr/sbin/clang++   -I/home/bstaletic/Temp/sparsehash-c11/. -I/home/bstaletic/Temp/sparsehash-c11/tests/.   -std=c++11 -Wall -Wextra -Wpedantic -Wno-missing-field-initializers -O3 -DNDEBUG -O3   -o CMakeFiles/sparsehash_unittests.dir/testmain.cc.o -c /home/bstaletic/Temp/sparsehash-c11/tests/testmain.cc",
  "file": "/home/bstaletic/Temp/sparsehash-c11/tests/testmain.cc"
},
{
  "directory": "/home/bstaletic/Temp/sparsehash-c11/build/tests",
  "command": "/usr/sbin/clang++   -I/home/bstaletic/Temp/sparsehash-c11/. -I/home/bstaletic/Temp/sparsehash-c11/tests/.   -std=c++11 -Wall -Wextra -Wpedantic -Wno-missing-field-initializers -O3 -DNDEBUG -O3   -o CMakeFiles/sparsehash_unittests.dir/sparsetable_unittests.cc.o -c /home/bstaletic/Temp/sparsehash-c11/tests/sparsetable_unittests.cc",
  "file": "/home/bstaletic/Temp/sparsehash-c11/tests/sparsetable_unittests.cc"
},
{
  "directory": "/home/bstaletic/Temp/sparsehash-c11/build/tests",
  "command": "/usr/sbin/clang++   -I/home/bstaletic/Temp/sparsehash-c11/. -I/home/bstaletic/Temp/sparsehash-c11/tests/.   -std=c++11 -Wall -Wextra -Wpedantic -Wno-missing-field-initializers -O3 -DNDEBUG -O3   -o CMakeFiles/sparsehash_unittests.dir/simple_unittests.cc.o -c /home/bstaletic/Temp/sparsehash-c11/tests/simple_unittests.cc",
  "file": "/home/bstaletic/Temp/sparsehash-c11/tests/simple_unittests.cc"
},
{
  "directory": "/home/bstaletic/Temp/sparsehash-c11/build/tests",
  "command": "/usr/sbin/clang++   -I/home/bstaletic/Temp/sparsehash-c11/. -I/home/bstaletic/Temp/sparsehash-c11/tests/.   -std=c++11 -Wall -Wextra -Wpedantic -Wno-missing-field-initializers -O3 -DNDEBUG -O3   -o CMakeFiles/sparsehash_unittests.dir/hashtable_unittests.cc.o -c /home/bstaletic/Temp/sparsehash-c11/tests/hashtable_unittests.cc",
  "file": "/home/bstaletic/Temp/sparsehash-c11/tests/hashtable_unittests.cc"
},
{
  "directory": "/home/bstaletic/Temp/sparsehash-c11/build/tests",
  "command": "/usr/sbin/clang++   -I/home/bstaletic/Temp/sparsehash-c11/. -I/home/bstaletic/Temp/sparsehash-c11/tests/.   -std=c++11 -Wall -Wextra -Wpedantic -Wno-missing-field-initializers -O3 -DNDEBUG -O3   -o CMakeFiles/sparsehash_unittests.dir/hashtable_c11_unittests.cc.o -c /home/bstaletic/Temp/sparsehash-c11/tests/hashtable_c11_unittests.cc",
  "file": "/home/bstaletic/Temp/sparsehash-c11/tests/hashtable_c11_unittests.cc"
},
{
  "directory": "/home/bstaletic/Temp/sparsehash-c11/build/tests",
  "command": "/usr/sbin/clang++   -I/home/bstaletic/Temp/sparsehash-c11/. -I/home/bstaletic/Temp/sparsehash-c11/tests/.   -std=c++11 -Wall -Wextra -Wpedantic -Wno-missing-field-initializers -O3 -DNDEBUG -O3   -o CMakeFiles/sparsehash_unittests.dir/fixture_unittests.cc.o -c /home/bstaletic/Temp/sparsehash-c11/tests/fixture_unittests.cc",
  "file": "/home/bstaletic/Temp/sparsehash-c11/tests/fixture_unittests.cc"
},
{
  "directory": "/home/bstaletic/Temp/sparsehash-c11/build/tests",
  "command": "/usr/sbin/clang++   -I/home/bstaletic/Temp/sparsehash-c11/. -I/home/bstaletic/Temp/sparsehash-c11/tests/.   -std=c++11 -Wall -Wextra -Wpedantic -Wno-missing-field-initializers -O3 -DNDEBUG -O3   -o CMakeFiles/sparsehash_unittests.dir/allocator_unittests.cc.o -c /home/bstaletic/Temp/sparsehash-c11/tests/allocator_unittests.cc",
  "file": "/home/bstaletic/Temp/sparsehash-c11/tests/allocator_unittests.cc"
},
{
  "directory": "/home/bstaletic/Temp/sparsehash-c11/build/tests/gtest",
  "command": "/usr/sbin/clang++   -I/home/bstaletic/Temp/sparsehash-c11/. -I/home/bstaletic/Temp/sparsehash-c11/tests/.   -std=c++11 -Wall -Wextra -Wpedantic -Wno-missing-field-initializers -O3 -DNDEBUG -O3   -o CMakeFiles/gtest.dir/gmock-gtest-all.cc.o -c /home/bstaletic/Temp/sparsehash-c11/tests/gtest/gmock-gtest-all.cc",
  "file": "/home/bstaletic/Temp/sparsehash-c11/tests/gtest/gmock-gtest-all.cc"
}
]
```

**Comment by sam-mccall** [comment]
Not clear to me whether files without extension that aren't in the CDB should be parsed with a fallback command (probably `-x objective-c++`) or ignored somehow, assuming they're plain text.

Regardless, the current behavior is clearly broken.

**Comment by bstaletic** [comment]
Ycmd has been blindly setting `-xc++` for those files and it has worked for us. Ignoring isn't a good option, because it would stop users (including me) from navigating inside STL headers.

**Comment by sam-mccall** [comment]
Fixed in r360153.

The change is that:
 - when there is a compilation database governing a file (found in ancestor dir or via flag)
 - and the file is not in the compilation database
 - and the file has no recognized extension
 - and we heuristically match it against some other file in the CDB, and use its compile command
then we'll add the `-x` flag from the other file (using the -header variant)

The case where there is no governing database, or we fail to heuristically match it, should have been working already, using the fallback command with `-x objective-c++`. The case where the extension is recognized should have already been adding `-x` where appropriate.

**Comment by abbaswasim** [comment]
I have similar issue but with files having a `.inl` extension. I am getting the same `invalid AST` message from `lsp`. I am pretty sure thats because `clangd` can't find flags and isn't compiling the file. I have been searching for a solution to this but haven't found anything yet. I have tried adding flags via `compile_flags.txt` but I think `clangd` is not considering `.inl` a valid source file. 
Sorry lots of speculation here but I hope someone more knowledgeable would know more.

**Comment by bstaletic** [comment]
Try adding an explicit language flag (`-x<language>`).

**Comment by abbaswasim** [comment]
I am not invoking `clangd` manually so can't test this directly but since I am using `lsp-mode` I have tried `(setq lsp-clients-clangd-args '("-j=4" "-background-index" "-log=error" "-xc++"))` and this isn't working. Is that what you meant?

**Comment by kadircet** [comment]
`.inl` is not considered a cpp extension by default on clang, you can find the list in https://github.com/llvm/llvm-project/blob/master/clang/lib/Driver/Types.cpp#L194 and send a quick patch explaining your reasoning why it should be considered as a cpp file. Then it will automatically get picked up by most of clang-based tools including clangd.

Otherwise you need to specify it manually within your compile commands container( compile_commands.json/compile_flags.txt/ycm_extra_conf). Make sure `-xc++` flag is stated before file name otherwise it has no effect.

**Comment by abbaswasim** [comment]
Thanks for the link I looked into this and for now I can get away with using `.hh` instead. I am trying to gather some reasoning for `.inl` and will submit an MR if I can justify it to myself. One of the problems with using `.inl` the way I am is that these are not complete translation unit so even if `clang` recognises it, it doesn't solve the problem. `clangd` still can't build it. So I had to include `.hpp` in `.inl` which is gross. By using `.hh` mean it comes early in the company suggestions which is annoying me :). `.inl` doesn't have that issue.
